File: projects/server/src/interactors/api/envelope.py
from typing import Any


def ok(data: Any, meta: dict[str, int] | None = None) -> dict[str, Any]:
    body: dict[str, Any] = {"success": True, "data": data, "error": None}
    if meta is not None:
        body["meta"] = meta
    return body


def ok_list(items: list[Any], total: int, page_size: int, page_number: int) -> dict[str, Any]:
    return ok(items, {"total": total, "page_size": page_size, "page_number": page_number})


def fail(message: str) -> dict[str, Any]:
    return {"success": False, "data": None, "error": message}

File: projects/ui/src/mocks/envelope.ts
import { HttpResponse } from "msw";

/** The API's envelope, so handlers and tests never hand-assemble it. */
export const ok = <T>(data: T) => HttpResponse.json({ success: true, data, error: null });

export const okList = <T>(items: T[]) =>
  HttpResponse.json({
    success: true,
    data: items,
    error: null,
    meta: { total: items.length, page_size: 50, page_number: 1 },
  });

export const created = <T>(data: T) =>
  HttpResponse.json({ success: true, data, error: null }, { status: 201 });

export const failure = (status: number, error: string) =>
  HttpResponse.json({ success: false, data: null, error }, { status });
